### PR TITLE
feat: add sourcepawn parser and queries

### PR DESCRIPTION
**Pull request by Sarrus1**
This PR adds a tree-sitter parser and queries for the [Sourcepawn programming language](https://wiki.alliedmods.net/Introduction_to_SourcePawn_1.7).

### CODE REVIEW

**Comment by clason** [comment]
Out of interest: There's already https://github.com/nilshelmig/tree-sitter-sourcepawn; can you comment on the differences in your parser?

**Comment by clason** [comment]
Also, be aware that `master` is frozen while we work on some query refactors (new capture naming scheme, formatting) on `main`, so this may have to wait a while (and then be adapted).

**Comment by lucario387** [comment]
> Out of interest: There's already https://github.com/nilshelmig/tree-sitter-sourcepawn; can you comment on the differences in your parser?

Looks to be a maintained fork of the parser

File: queries/sourcepawn/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,255 @@
+; Assume all-caps names are constants
+((identifier) @constant
+ (#match? @constant "^[A-Z][A-Z\\d_]+$'"))
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:3]:
```suggestion
 (#lua-match? @constant "^[A-Z][A-Z0-9_]+$"))
```

File: queries/sourcepawn/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+(preproc_endif) @preproc
+(preproc_endinput) @preproc
+(preproc_define) @define
+(preproc_define
+  name: (identifier) @constant)
+(preproc_undefine) @define
+(preproc_undefine
+  name: (identifier) @constant)
+(preproc_error) @function.macro ; Wrong color?
+(preproc_warning) @function.macro ; Wrong color?
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:59]:
```suggestion
(preproc_error) @function.macro
(preproc_warning) @function.macro
```

File: queries/sourcepawn/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+; https://github.com/alliedmodders/sourcemod/blob/5c0ae11a4619e9cba93478683c7737253ea93ba6/plugins/include/handles.inc#L78
+(hardcoded_symbol) @variable.builtin
+
+; Comments
+(comment) @comment
+
+; General
+(parameter_declaration
+  defaultValue: (identifier) @constant)
+(fixed_dimension) @punctuation.bracket ; the [3] in var[3]
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:84]: The `3` should be `@number`.

File: queries/sourcepawn/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+(float_literal) @float
+(string_literal) @string
+(array_literal) @punctuation.bracket
+(concatenated_string) @punctuation.delimiter ; Dots in string-literal concat--Doesn't seem to work?
+[
+  (bool_literal)
+  (null)
+] @constant.builtin
+((identifier) @constant
+  (#match? @constant "INVALID_HANDLE"))
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:208]:
```suggestion
  (#eq? @constant "INVALID_HANDLE"))
```

File: queries/sourcepawn/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+; Comment specialisations (must be after comment)
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:210]: Remove these and add a comment injection:

```clojure
((comment) @injection.content
 (#set! injection.language "comment"))
 ```

File: queries/sourcepawn/highlights.scm [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+  "stock"
+  "struct"
+  "switch"
+  "typedef"
+  "typeset"
+  "void"
+  "while"
+] @keyword
+
+(identifier) @variable
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:255]: Should be on top.

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

**Comment by clason** [comment]
> > Out of interest: There's already https://github.com/nilshelmig/tree-sitter-sourcepawn; can you comment on the differences in your parser?
> 
> Looks to be a maintained fork of the parser

Upstream doesn't look abandoned, tho?

File: queries/sourcepawn/highlights.scm (hunk 3 above)
Comment (Sarrus1) [reply queries/sourcepawn/highlights.scm]: The comment was a leftover and no longer relevant, I removed it

File: queries/sourcepawn/highlights.scm (hunk 5 above)
Comment (Sarrus1) [reply queries/sourcepawn/highlights.scm]: There is already a `jsdoc` injection [here](https://github.com/Sarrus1/nvim-treesitter/blob/17704fc1b710d0b18a1e7c011ba6b0de303abf4e/queries/sourcepawn/injections.scm)

I removed the problematic query in the `highlights.scm` file and kept the `jsdoc` injection 👍

**Comment by Sarrus1** [comment]
> Out of interest: There's already https://github.com/nilshelmig/tree-sitter-sourcepawn; can you comment on the differences in your parser?

@nilshelmig was the first to write a tree-sitter grammar for SourcePawn. It was a direct adaptation of the official c grammar.
I forked it and improved it when I started working on [SPFormat](https://github.com/Sarrus1/spformat) and later on [sourcepawn-vscode](https://github.com/Sarrus1/sourcepawn-vscode).
I open a PR on @nilshelmig's repo everytime I make a change, but they tend to take a while to get merged...
I made the past 125 commits, out of the ~230 of the repo.
Since PR take a long time to get merged I thought I would add myself as the maintainer of the queries here.

TL;DR:
Upstream is not officially abandoned, but PR take months to get merged, as the original maintainer probably does not look at his notifications.

> Also, be aware that master is frozen while we work on some query refactors (new capture naming scheme, formatting) on main, so this may have to wait a while (and then be adapted).

Thanks for the head ups :)

**Comment by clason** [comment]
Don't forget to run `./scripts/update-readme.lua`.

File: queries/sourcepawn/highlights.scm (hunk 5 above)
Comment (ObserverOfTime) [reply queries/sourcepawn/highlights.scm]: jsdoc does not include todo highlights.

File: runtime/queries/sourcepawn/highlights.scm [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+  "set"
+  "static"
+  "stock"
+  "struct"
+  "switch"
+  "typedef"
+  "typeset"
+  "void"
+  "while"
+] @keyword
```
Comment (ObserverOfTime) [inline runtime/queries/sourcepawn/highlights.scm:245]: Some of the keywords should be more specific (e.g. `@conditional` / `@repeat`).

**Comment by clason** [comment]
Sorry; you need to rebase on `main` (and run `./scripts/format-queries.lua runtime/queries/sourcepawn`).

**Comment by clason** [comment]
I would do it for you, but I can't since you opened your PR straight from your `master` (which is one reason why you should never do that; always make a PR from a feature branch).

**Comment by Sarrus1** [comment]
> I would do it for you, but I can't since you opened your PR straight from your `master` (which is one reason why you should never do that; always make a PR from a feature branch).

Sorry about that

> run `./scripts/format-queries.lua runtime/queries/sourcepawn.`

I have tried but I get:
```
E5113: Error while calling lua chunk: ./scripts/format-queries.lua:24: attempt to call field 'list_contains' (a nil value)
stack traceback:
	./scripts/format-queries.lua:24: in function 'handler'
	...im/0.9.1/share/nvim/runtime/lua/vim/treesitter/query.lua:561: in function 'match_preds'
	...im/0.9.1/share/nvim/runtime/lua/vim/treesitter/query.lua:650: in function '(for generator)'
	./scripts/format-queries.lua:385: in function 'format'
	./scripts/format-queries.lua:400: in main chunk%     
```

> Don't forget to run `./scripts/update-readme.lua`.
I also get an error:
```
E5113: Error while calling lua chunk: ./scripts/update-readme.lua:53: attempt to index field 'uv' (a nil value)
stack traceback:
	./scripts/update-readme.lua:53: in main chunk% 
```

i have `NVIM v0.9.1` installed. Sorry I'm not good with Lua and this is my first contribution to the neovim ecosystem, can you please give me some pointers as to how to resolve this?

Should I open another PR from a feature branch?

**Comment by clason** [comment]
Nvim 0.9.1 is outdated; please install 0.9.5. (The formatter _should_ work with that, but if not, just download a nightly appimage from the Github releases page and run that locally.)

**Comment by clason** [comment]
> Should I open another PR from a feature branch?

You can do that as well if you prefer. (Or make the changes manually, of course.)

**Comment by Sarrus1** [comment]
> Nvim 0.9.1 is outdated; please install 0.9.5. (~~The formatter _should_ work with that, but if not, ~~ The scripts on `main` indeed rely on Nvim nightly; just download a nightly appimage from the Github releases page and run that locally with `<path to nvim-appimage> -l <script>`.)

I ended up changing the `nvim` path in the lua shebang headers, that worked 👍 

Everything should be in order now, thank your for your patience!

**Comment by Sarrus1** [comment]
Oh I thought there would be a confirmation prompt for this, sorry for the ping...

File: plugin/filetypes.lua [hunk 8]
```diff
@@ -29,6 +29,7 @@ local filetypes = {
   ssh_config = { 'sshconfig' },
   starlark = { 'bzl' },
   surface = { 'sface' },
+  sourcepawn = { 'sp', 'inc' },
```
Comment (ObserverOfTime) [inline plugin/filetypes.lua:32]: This is for filetypes, not extensions.
If sourcepawn does not currently have a filetype, it should be added to Vim.

```suggestion
```

File: runtime/queries/sourcepawn/highlights.scm [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+(char_literal) @character
+
+(float_literal) @float
+
+(string_literal) @string
+
+(array_literal) @punctuation.bracket
+
+(concatenated_string) @punctuation.delimiter ; Dots in string-literal concat--Doesn't seem to work?
+
```
Comment (ObserverOfTime) [inline runtime/queries/sourcepawn/highlights.scm:275]: If it doesn't work, it should either be fixed or removed.

```suggestion
```

File: runtime/queries/sourcepawn/highlights.scm [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+[
+  (bool_literal)
+  (null)
+] @constant.builtin
```
Comment (ObserverOfTime) [inline runtime/queries/sourcepawn/highlights.scm:279]:
```suggestion
(bool_literal) @boolean

(null) @constant.builtin
```

File: runtime/queries/sourcepawn/highlights.scm [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+[
+  "break"
+  "continue"
+  "return"
+] @keyword.return
+
+[
+  "if"
+  "else"
+] @conditional
```
Comment (ObserverOfTime) [inline runtime/queries/sourcepawn/highlights.scm:294]: Should also include "switch", "case", "default".

File: runtime/queries/sourcepawn/highlights.scm [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+  (null)
+] @constant.builtin
+
+((identifier) @constant
+  (#eq? @constant "INVALID_HANDLE"))
+
+; Keywords
+[
+  "break"
+  "continue"
```
Comment (ObserverOfTime) [inline runtime/queries/sourcepawn/highlights.scm:287]: These should be in `@repeat`.

File: runtime/queries/sourcepawn/highlights.scm [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+  "public"
+  "set"
+  "static"
+  "stock"
+  "struct"
+  "switch"
+  "typedef"
+  "typeset"
+  "void"
+] @keyword
```
Comment (ObserverOfTime) [inline runtime/queries/sourcepawn/highlights.scm:326]: "\_\_nullable\_\_", "const", "native" could be `@type.qualifier`.

**Review by ObserverOfTime — COMMENTED** [review]
Since the PR targets `main` now, it should use the new captures (#5831).

File: runtime/queries/sourcepawn/highlights.scm (hunk 9 above)
Comment (Sarrus1) [reply runtime/queries/sourcepawn/highlights.scm]: Leftover from a previous commit, it works 👍

**Comment by clason** [comment]
(Test failure is just the regularly scheduled githubocalypse.)

**Comment by clason** [comment]
Actually, the remaining one isn't; please run `./scripts/format-queries runtime/queries/sourcepawn`.

**Comment by clason** [comment]
Yes, and drop the rebased commits from `main`. (Or wait until we have backported everything and are ready to merge this to `master`.)

**Comment by Sarrus1** [comment]
Should I rebase on master and target master?

**Comment by clason** [comment]
Please wait for https://github.com/nvim-treesitter/nvim-treesitter/pull/5895 to be merged (hopefully before the end of the week).

**Comment by clason** [comment]
The backport PR has been merged; please **rebase on master** and note 1. the new list of valid capture names and 2. the mandatory query format (run `./scripts/format-queries.lua` on the new queries).

**Comment by clason** [comment]
ping @Sarrus1 

(I cannot rebase for you since you didn't open this PR from a branch.)

**Comment by clason** [comment]
@Sarrus1 ping -- are you still interested in finishing this PR (and maintaining the queries)?

**Comment by Sarrus1** [comment]
Exam season has just ended, I will be able to finish the PR the coming week! And to continue maintain it in the future 👍

**Comment by Sarrus1** [comment]
Just a precision regarding the parser's repo status: I was recently added as a maintainer on the original repo, therefore I have switched it back to its url in `parsers.lua`.

File: README.md [hunk 14]
```diff
@@ -373,6 +373,7 @@ We are looking for maintainers to add more parsers and to write query files for
 - [x] [solidity](https://github.com/JoranHonig/tree-sitter-solidity) (maintained by @amaanq)
 - [x] [soql](https://github.com/aheber/tree-sitter-sfapex) (maintained by @aheber)
 - [x] [sosl](https://github.com/aheber/tree-sitter-sfapex) (maintained by @aheber)
+- [x] [sourcepawn](https://github.com/nilshelmig/tree-sitter-sourcepawn) | community | `H  JL` | | | @Sarrus1
```
Comment (lucario387) [inline README.md:376]: .

File: queries/sourcepawn/highlights.scm [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+
+(preproc_if) @keyword.directive
+
+(preproc_else) @keyword.directive
+
+(preproc_elseif) @keyword.directive
+
+(preproc_endif) @keyword.directive
+
+(preproc_endinput) @keyword.directive
```
[the remark this answers is not in the document]
Comment (lucario387) [reply queries/sourcepawn/highlights.scm:74]: Similar to other places with the same capture

File: queries/sourcepawn/highlights.scm [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+(enum_struct_field
+  name: (identifier) @variable.member)
+
+(enum_struct_method
+  name: (identifier) @function.method)
+
+; Non-type Keywords
+(variable_storage_class) @keyword.storage
+
+(visibility) @keyword.storage
```
Comment (lucario387) [inline queries/sourcepawn/highlights.scm:162]: You can group these two

File: queries/sourcepawn/highlights.scm [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+(preproc_define
+  name: (identifier) @constant)
+
+(preproc_undefine
+  name: (identifier) @constant)
+
+[
+  (preproc_error)
+  (preproc_warning)
+] @keyword.directive.error
```
Comment (lucario387) [inline queries/sourcepawn/highlights.scm:88]: I remember after grouping this capture changed. Unsure why

Comment (Sarrus1) [reply queries/sourcepawn/highlights.scm]: It makes more sense for them to be `@keyword.directive` (I removed the `.error` as this is incorrect, sorry about that).

File: queries/sourcepawn/highlights.scm [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+    (field_access
+      field: (identifier) @function.method.call))
+
+; Types
+(builtin_type) @type.builtin
+
+(type
+  (identifier) @type)
+
+(any_type) @type
```
Comment (amaanq) [inline queries/sourcepawn/highlights.scm:36]: maybe `@type.builtin`? or is this like a user type, if so then nvm

Comment (Sarrus1) [reply queries/sourcepawn/highlights.scm]: You are correct, I've pushed the changes, thanks!

**Review by amaanq — APPROVED** [review]

**Comment by amaanq** [comment]
@ObserverOfTime

File: queries/sourcepawn/highlights.scm [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+(sizeof_expression) @function.macro
+
+[
+  (this)
+  ; https://github.com/alliedmodders/sourcemod/blob/5c0ae11a4619e9cba93478683c7737253ea93ba6/plugins/include/handles.inc#L78
+  (hardcoded_symbol)
+] @variable.builtin
+
+; Comments
+(comment) @comment
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:98]:
```suggestion
(comment) @comment @spell
```

File: queries/sourcepawn/highlights.scm [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+
+; General
+(parameter_declaration
+  defaultValue: (identifier) @constant)
+
+[
+  (fixed_dimension)
+  (dimension)
+  (array_indexed_access)
+] @punctuation.bracket
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:108]: Do these include the index?

File: queries/sourcepawn/highlights.scm [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+
+(enum_struct_method
+  name: (identifier) @function.method)
+
+; Non-type Keywords
+[
+  (variable_storage_class)
+  (visibility)
+  (variable_storage_class)
+] @keyword.storage
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:160]:
```suggestion
(variable_storage_class) @keyword.storage

(visibility) @type.qualifier
```

File: queries/sourcepawn/highlights.scm [hunk 22]
```diff
@@ -0,0 +1,306 @@
...
+(array_literal) @punctuation.bracket
+
+(concatenated_string) @punctuation.delimiter
+
+(bool_literal) @boolean
+
+(null) @constant.builtin
+
+((identifier) @constant
+  (#eq? @constant "INVALID_HANDLE"))
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:261]:
```suggestion
((identifier) @constant.builtin
  (#eq? @constant.builtin "INVALID_HANDLE"))
```

File: queries/sourcepawn/highlights.scm [hunk 23]
```diff
@@ -0,0 +1,306 @@
...
+(struct_declaration
+  name: (identifier) @variable.builtin)
+
+; Typedef/Typedef
+[
+  (typedef)
+  (functag)
+  (funcenum)
+  (typeset)
+] @type.definition
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:227]: Note that `@type.definition` is meant to be used for the type alias and not the keyword that defines it.

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

File: queries/sourcepawn/highlights.scm (hunk 20 above)
Comment (Sarrus1) [reply queries/sourcepawn/highlights.scm:108]: Only `fixed_dimension`.

Comment (ObserverOfTime) [reply queries/sourcepawn/highlights.scm:108]: Is it highlighted as a number below?

Comment (Sarrus1) [reply queries/sourcepawn/highlights.scm:106]: Yep, just rechecked, it looks good in practice 👍

File: queries/sourcepawn/highlights.scm (hunk 23 above)
Comment (Sarrus1) [reply queries/sourcepawn/highlights.scm]: Ah in that case, this was incorrect, it has been removed now.

File: queries/sourcepawn/highlights.scm [hunk 24]
```diff
@@ -0,0 +1,302 @@
...
+  (rest_operator)
+] @operator
+
+; public Plugin myinfo
+(struct_declaration
+  name: (identifier) @variable.builtin)
+
+; Typedef/Typedef
+(typedef
+  name: (identifier) @type)
```
Comment (ObserverOfTime) [inline queries/sourcepawn/highlights.scm:215]:
```suggestion
  name: (identifier) @type.definition)
```

Comment (Sarrus1) [reply queries/sourcepawn/highlights.scm:215]: This is not a typedef like c, it looks like this in sourcepawn:

```sourcepawn
typedef SrvCmd = function Action (int args);
```

This is to specify the prototype of a function, I don't think this should use `@type.definition`?

**Review by ObserverOfTime — APPROVED** [review]

**Comment by Sarrus1** [comment]
Should I squash everything?

**Comment by lucario387** [comment]
It's fine :)
We can squash it

**Comment by Sarrus1** [comment]
Thank you @lucario387, @clason, @ObserverOfTime for your patience and helping me to get this merged!
